pkg/reconciler/workspace/controller.go

```diff
@@ -0,0 +1,385 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/rand"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/clusters"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+
+	tenancyv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/tenancy/v1alpha1"
+	kcpclient "github.com/kcp-dev/kcp/pkg/client/clientset/versioned"
+	tenancyinformer "github.com/kcp-dev/kcp/pkg/client/informers/externalversions/tenancy/v1alpha1"
+	tenancylister "github.com/kcp-dev/kcp/pkg/client/listers/tenancy/v1alpha1"
+	"github.com/kcp-dev/kcp/pkg/util/errors"
+)
+
+const targetShardName = "targetShard"
+const unschedulable = "unschedulable"
+
+func NewController(
+	kcpClient kcpclient.ClusterInterface,
+	workspaceInformer tenancyinformer.WorkspaceInformer,
+	workspaceShardInformer tenancyinformer.WorkspaceShardInformer,
+) (*Controller, error) {
+	queue := workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter())
+
+	c := &Controller{
+		queue:                 queue,
+		kcpClient:             kcpClient,
+		workspaceIndexer:      workspaceInformer.Informer().GetIndexer(),
+		workspaceLister:       workspaceInformer.Lister(),
+		workspaceShardIndexer: workspaceShardInformer.Informer().GetIndexer(),
+		workspaceShardLister:  workspaceShardInformer.Lister(),
+	}
+
+	workspaceInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    func(obj interface{}) { c.enqueue(obj) },
+		UpdateFunc: func(_, obj interface{}) { c.enqueue(obj) },
+	})
+	if err := c.workspaceIndexer.AddIndexers(map[string]cache.IndexFunc{
+		targetShardName: func(obj interface{}) ([]string, error) {
+			if workspace, ok := obj.(*tenancyv1alpha1.Workspace); ok {
+				return []string{workspace.Spec.TargetShard}, nil
+			}
+			return []string{}, nil
+		},
+		unschedulable: func(obj interface{}) ([]string, error) {
+			if workspace, ok := obj.(*tenancyv1alpha1.Workspace); ok {
+				if IsWorkspaceUnschedulable(workspace) {
+					return []string{"true"}, nil
+				}
+			}
+			return []string{}, nil
+		},
+	}); err != nil {
+		return nil, fmt.Errorf("failed to add indexer for Workspace: %w", err)
+	}
+
+	workspaceShardInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    func(obj interface{}) { c.enqueueAddedShard(obj) },
+		DeleteFunc: func(obj interface{}) { c.enqueueDeletedShard(obj) },
+	})
+
+	return c, nil
+}
+
+type Controller struct {
+	queue workqueue.RateLimitingInterface
+
+	kcpClient        kcpclient.ClusterInterface
+	workspaceIndexer cache.Indexer
+	workspaceLister  tenancylister.WorkspaceLister
+
+	workspaceShardIndexer cache.Indexer
+	workspaceShardLister  tenancylister.WorkspaceShardLister
+}
+
+func (c *Controller) enqueue(obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+	klog.Infof("queueing workspace %q", key)
```

Comment:
general question: shouldn't we move to structured logging right away?

Reply:
I'm all for it - how, using `klog`, is that supported?

Reply:
I reviewed some PRs in kube, klog.InfoS I believe. Never done that before. But if we start fresh anyway, do it right and enjoy beautiful json log in the future.